### PR TITLE
AudioScheduledSourceNode

### LINKED ISSUE #21: Implement support for AudioScheduledSourceNode start and stop methods
- [AudioScheduledSourceNode.start](https://webaudio.github.io/web-audio-api/#dom-audioscheduledsourcenode-start)
- [AudioScheduledSourceNode.stop](https://webaudio.github.io/web-audio-api/#dom-audioscheduledsourcenode-stop)

### PR DESCRIPTION
**Pull request by ferjm**
Fixes  #21.

### CODE REVIEW

**Comment by Manishearth** [comment]
Overall looks good.




Also, is it possible to schedule multiple events? If so we'd need to store these as a list of events (in a smallvec probably) and check them each time.

**Comment by ferjm** [comment]
> Also, is it possible to schedule multiple events? If so we'd need to store these as a list of events (in a smallvec probably) and check them each time

Good point! Yes, I believe it's possible.

**Comment by Manishearth** [comment]
Also, we should ensure that these events only need to be checked per-block
(not frame) when we handle the timings.

Otherwise we need to make this part of an updateparameters step.

AudioParams need something like this too, so if you want we can leave it as
is as a k-type schedule till I add that support.

**Comment by Manishearth** [comment]
Given that we now have ticks, can we internally use ticks for this? We can convert time floats to ticks at the user API boundary

**Comment by ferjm** [comment]
@Manishearth I think this is ready for a final review. I checked the spec and found that we cannot schedule more than one start or stop times, so my believes [here](https://github.com/servo/media/pull/31#issuecomment-392723901) where incorrect.

File: src/audio/gain_node.rs [hunk 1]
```diff
@@ -56,11 +59,6 @@ impl AudioNodeEngine for GainNode {
         }
         inputs
     }
-    fn message(&mut self, msg: AudioNodeMessage, sample_rate: f32) {
-        match msg {
-            AudioNodeMessage::SetAudioParamEvent(event) => {
-                self.gain.insert_event(event.to_event(sample_rate))
```
Comment (Manishearth) [inline src/audio/gain_node.rs]: This reverts the AudioParam changes for gainnode

File: src/audio/node.rs [hunk 2]
```diff
@@ -41,20 +41,26 @@ impl BlockInfo {
...
-
-    }
+    fn message(&mut self, _: AudioNodeMessage, _sample_rate: f32) {}
 }
 
-
 pub enum AudioNodeMessage {
-    SetAudioParamEvent(UserAutomationEvent)
+    OscillatorNode(OscillatorNodeMessage),
+    GainNode(GainNodeMessage),
```
Comment (Manishearth) [inline src/audio/node.rs]: I think we should be able to send all kinds of messages to all nodes, and nodes filter what they can actually handle.

Also, this reverts the AudioParam changes.

So you have SetAudioParam, Start, and Stop.

File: src/audio/gain_node.rs (hunk 1 above)
Comment (ferjm) [reply src/audio/gain_node.rs]: They are moved [here](https://github.com/servo/media/pull/31/files/3748b6b36b0ee84ed1dc7ba7e0655e41b84d88f2#diff-ab2867bdbfa223725105fd3200838b45R36)

Comment (Manishearth) [reply src/audio/gain_node.rs]: Oh, I see what you did.

(sorry, reviewing on a phone)

I guess this is ok. A lot of the messages are common between nodes so I thought we'd have common messages. But this works.

**Pull request merged**